exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,419 @@
+            let height:u64 = CoreSchema::new(snapshot).height().into();
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::after_commit_id(),
+                    JavaType::Primitive(Primitive::Void),
+                    &[
+                        JValue::from(view_handle),
+                        JValue::from(validator_id),
+                        JValue::from(height as i64),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+
+    // TODO: consider connecting api during the service start due to warning:
+    // "It is a temporary method which retains the existing `RustRuntime` code"
+    fn api_endpoints(&self, context: &ApiContext) -> Vec<(String, ServiceApiBuilder)> {
+        let started_ids = Vec::<i32>::new();
+        let node = NodeContext::new(self.exec.clone(), context.clone());
+
+        unwrap_jni(self.exec.with_attached(|env| {
+            let node_handle = to_handle(node);
+            let ids_array = env.new_int_array(started_ids.capacity() as i32)?;
+            env.set_int_array_region(ids_array, 0, &started_ids)?;
+            let service_ids = JObject::from(ids_array);
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::connect_apis_id(),
+                    JavaType::Primitive(Primitive::Void),
+                    &[
+                        JValue::from(service_ids),
+                        JValue::from(node_handle),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+
+        Vec::<(String, ServiceApiBuilder)>::new()
+    }
+}
+
+impl ToString for JavaArtifactId {
+    fn to_string(&self) -> String {
+        self.0.to_owned()
+    }
+}
+
+impl ServiceStateHashes {
+    fn to_hashes(&self) -> (InstanceId, Vec<Hash>) {
+        let hashes = self.state_hashes
+            .iter()
+            .map(|bytes| Hash::from_bytes(bytes.into()).unwrap())
+            .collect();
+        (self.instance_id, hashes)
+    }
+}
+
+impl ServiceRuntimeStateHashes {
+    fn runtime(&self) -> Vec<Hash> {
+        self.runtime_state_hashes
+            .iter()
+            .map(|bytes| Hash::from_bytes(bytes.into()).unwrap())
+            .collect()
+    }
+
+    fn instances(&self) -> Vec<(InstanceId, Vec<Hash>)> {
+        self.service_state_hashes
+            .iter()
+            .map(ServiceStateHashes::to_hashes)
+            .collect()
+    }
+}
+
+impl From<ServiceRuntimeStateHashes> for StateHashAggregator {
+    fn from(value: ServiceRuntimeStateHashes) -> Self {
+        StateHashAggregator {
+            runtime: value.runtime(),
+            instances: value.instances(),
+        }
+    }
+}
```

Comment:
```suggestion
}

From<JavaRuntimeProxy> for (u32, Box<dyn Runtime>)
    fn from(r: JavaRuntimeProxy) -> Self {
        (JavaRuntimeProxy::RUNTIME_ID as u32, Box::new(r))
    }
```

Reply:
Sorry - somehow I lost parts of the code when copied it and didn't noticed that 😟. Now it's correct.